Add support for custom error message

joeljames opened the pull request:
This PR adds the ability to add custom error messages on serializer fields. Click  https://github.com/localmed/pyserializer/issues/12 for more info on the issue.

@petebrowne @reneric whenever you get a chance can you guys take a look at this PR?
And if every thing looks good can you guys push up a new release after merging it to `localmed/pyserializer` master branch. I have already bumped the version number. This release should be `0.8.0`. You will also have to build a new version of docs which gets published to [Read the Docs](http://pyserializer.readthedocs.io/en/latest/) as I have made changes to the docs.

joeljames committed:
Add python 3.5 testing env

joeljames committed:
Add support for custom error messages on serializer fields

joeljames committed:
Some clean up

joeljames committed:
Update change log

joeljames committed:
Add documentation for custom error message on fields

joeljames committed:
Move custom error message to a separate module

joeljames committed:
Fix linting issues

joeljames committed:
Bump version: 0.7.0 → 0.8.0

joeljames commented on tests/functional_tests/pyserializer/custom_error_message_tests.py:
```diff
@@ -0,0 +1,46 @@
+
+        self.UserDeserializer = UserDeserializer
+
+    def test_sets_custom_error_message(self):
```
This is the unit test that tests the functionality of the custom error message.

joeljames commented on tox.ini:
```diff
@@ -1,5 +1,5 @@
 [tox]
-envlist = py27, py33, py34
+envlist = py27, py33, py34, py35
```
I also went ahead and added testing env for python 3.5

petebrowne commented on pyserializer/fields.py:
```diff
@@ -58,11 +60,24 @@ def __init__(self,
+                'type_label',
+                self.type_label
+            )
+            self.error_dict['message'] = error_messages['invalid']
```
If I'm following this correctly, it only seems to handle one type of error message (`invalid`). Wouldn't it be better to allow the user to override the error message per validation type? for example:

``` python
class UserDeserializer(Serializer):
  age = fields.IntegerField(
    validators=[MinValueValidator(18)],
    error_messages={
      'min_value': 'You must be at least 18 years old to signup',
    },
  )
```

joeljames replied:
Can you elaborate on what you mean by "override the error message per validation type". For example if you have a UserDeserializer with age field having a range validation(min and max limit on age). What should your expected error message look like? Are you suggesting the error messages should be passed in validator constructor and not the field constructor?
```
class UserDeserializer(Serializer):
  age = fields.IntegerField(
    validators=[MinValueValidator(18), MaxValueValidator(30)],
    error_messages={
      'min_value': 'You must be at least 18 years old to signup',
    },
  )
```

petebrowne replied:
For your example I see a couple of options for the API that would allow you to customize the error message for each validator:

```python
# Option 1
# Override the error messages from the validators at the field level
class UserDeserializer(Serializer):
  age = fields.IntegerField(
    validators=[MinValueValidator(18), MaxValueValidator(30)],
    error_messages={
      'min_value': 'You must be at least 18 years old to signup',
      'max_value': 'You are too old...',
    },
  )

# Option 2
# Customize the error messages at the validator level
class UserDeserializer(Serializer):
  # Custom error messages per validator
  age_1 = fields.IntegerField(
    validators=[
        MinValueValidator(18, error_message='You must be at least 18 years old to signup'),
        MaxValueValidator(30, error_message='You are too old...'),
    ],
  )
  # One error message if any validator fails:
  age_2 = fields.IntegerField(
    validators=[MinValueValidator(18), MaxValueValidator(30)],
    error_message='Your age is invalid',
  )
```

joeljames replied:
Yes, that makes sense. What would you prefer option 1 or option 2? I can update the PR based on it.

petebrowne replied:
It's a tough call haha. I could go either way... @reneric  what do you think

joeljames replied:
I think I prefer option 1, since thats a pattern I have seen in Django Rest Framework.

joeljames committed:
Adress review comments

joeljames commented:
@petebrowne I have addressed your comments. Please take a look at the unit tests `test_sets_custom_error_message` to verify if my implementation matches every thing you requested.

joeljames commented:
@petebrowne can you merge the PR in, and then merge it to localmed/pyserializer master branch and push out a release? I don't have push/pull/merge access to localmed/pyserialize anymore.

petebrowne commented:
Thanks @joeljames  👍
